plugin/pkg/auth/authenticator/password/keystone/keystone.go

```diff
@@ -45,18 +50,74 @@ func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(usernam
 		glog.Info("Failed: Starting openstack authenticate client")
 		return nil, false, errors.New("Failed to authenticate")
 	}
-
+//FIXME user_id, not user_name...
 	return &user.DefaultInfo{Name: username}, true, nil
 }
 
-// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
-func NewKeystoneAuthenticator(authURL string) (*KeystoneAuthenticator, error) {
-	if !strings.HasPrefix(authURL, "https") {
-		return nil, errors.New("Auth URL should be secure and start with https")
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	cfg, provider, err := loadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct { User struct { Id string }}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	return &user.DefaultInfo{Name: obj.Token.User.Id}, response.StatusCode == 200, nil
+}
+
+func loadConfig(configPath string) (openstackcp.Config, *gophercloud.ProviderClient, error) {
+	var cf *os.File
+	var err error
+	cf, err = os.Open(configPath)
+	if err != nil {
+		glog.Fatalf("Couldn't open configuration %s: %#v",
+			configPath, err)
 	}
-	if authURL == "" {
-		return nil, errors.New("Auth URL is empty")
+	defer cf.Close()
+	cfg, provider, err := openstackcp.ConfigToProvider(cf)
+	if err != nil {
+		return cfg, nil, err
 	}
+	if !strings.HasPrefix(cfg.Global.AuthUrl, "https") {
+		return cfg, nil, errors.New("Auth URL should be secure and start with https")
+	}
+	if cfg.Global.AuthUrl == "" {
+		return cfg, nil, errors.New("Auth URL is empty")
+	}
+	return cfg, provider, nil
+}
 
-	return &KeystoneAuthenticator{authURL}, nil
+// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
+func NewKeystoneAuthenticator(configPath string) (*KeystoneAuthenticator, error) {
```

Comment:
I think It is a good chance to change this return value to be `authenticator.Password`.